interface conversion: *pgadapter.Adapter is not persist.UpdatableAdapter: missing method UpdateFilteredPolicies

Hello, when running `UpdatePolicy` this error comes up.

```bash
interface conversion: *pgadapter.Adapter is not persist.UpdatableAdapter: missing method UpdateFilteredPolicies
```

my configuration:
```go
import (
	db "backend/postgres"
	"encoding/json"
	"log"

	pgadapter "github.com/casbin/casbin-pg-adapter"
	"github.com/casbin/casbin/v2"
	"github.com/casbin/casbin/v2/model"
	defaultrolemanager "github.com/casbin/casbin/v2/rbac/default-role-manager"
	"github.com/casbin/casbin/v2/util"
)

var Enforcer *casbin.Enforcer

func Init() error {
	a, err := pgadapter.NewAdapterByDB(db.Db, pgadapter.WithTableName("casbin"))
	if err != nil {
		return err
	}

	m, err := model.NewModelFromString(`
[request_definition]
r = sub, dom, obj, act

[policy_definition]
p = sub, dom, obj, act

[role_definition]
g = _, _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub, r.dom) && (r.dom == p.dom) && (r.dom == p.dom) && keyMatch2(r.obj, p.obj) && regexMatch(r.act, p.act)
`)

	if err != nil {
		log.Fatalf("error: model: %s", err)
	}

	enforcer, err := casbin.NewEnforcer(m, a)

	if err != nil {
		return err
	}

	r := enforcer.GetRoleManager()
	r.(*defaultrolemanager.RoleManager).AddDomainMatchingFunc("KeyMatch2", util.KeyMatch2)

	Enforcer = enforcer
	return AddDefaultPoliciesAndGroups()
}
```

the code that triggers the error:
```go
oldPolicy := []string{strings.ToLower(subject), strings.ToLower(domain), object, oldAction}
newPolicy := []string{strings.ToLower(subject), strings.ToLower(domain), object, action}

_, err := enforcer.Enforcer.UpdatePolicy(oldPolicy, newPolicy)
if err != nil {
	log.Println(err)
	return nil, err
}
```

It doesn't work in my setup.

This is how I create the initial policies (this function is called right after my Init() function in the original post):
```go
func AddDefaultPoliciesAndGroups() error {
	// p = sub, dom, obj, act
	policies := [][]string{
		{"admin", "domain1", "users", "allow"},
	}

	_, err := Enforcer.AddPolicies(policies)
	if err != nil {
		return err
	}

	groups := [][]string{
		{"admin@domain1.com", "admin", "domain1"},
	}

	_, err = Enforcer.AddNamedGroupingPolicies("g", groups)
	if err != nil {
		return err
	}

// I added this to test, and it still throws the same error
	_, err = Enforcer.UpdatePolicy([]string{"admin", "domain1", "users", "allow"}, []string{"admin", "domain1", "users", "deny"})
	if err != nil {
		return err
	}

	return nil
}
```

Maybe is it because I'm using domains? 

This is the full panic log:
```
panic: interface conversion: *pgadapter.Adapter is not persist.UpdatableAdapter: missing method UpdateFilteredPolicies

goroutine 1 [running]:
github.com/casbin/casbin/v2.(*Enforcer).updatePolicy(0xc0000c6800, 0x1088809, 0x1, 0x1088809, 0x1, 0xc00015d600, 0x4, 0x4, 0xc00015d640, 0x4, ...)
        /home/dearr/go/pkg/mod/github.com/casbin/casbin/v2@v2.30.0/internal_api.go:154 +0x459
github.com/casbin/casbin/v2.(*Enforcer).UpdateNamedPolicy(...)
        /home/dearr/go/pkg/mod/github.com/casbin/casbin/v2@v2.30.0/management_api.go:166
github.com/casbin/casbin/v2.(*Enforcer).UpdatePolicy(0xc0000c6800, 0xc00015d600, 0x4, 0x4, 0xc00015d640, 0x4, 0x4, 0x1, 0xc000511c00, 0x0)
        /home/dearr/go/pkg/mod/github.com/casbin/casbin/v2@v2.30.0/management_api.go:162 +0xa5
backend/enforcer.AddDefaultPoliciesAndGroups(0xf253a0, 0xc0000a4600)
        /home/dearr/Work/whistle/backend/enforcer/enforcer.go:97 +0xcf1
backend/enforcer.Init(0x0, 0x0)
        /home/dearr/Work/whistle/backend/enforcer/enforcer.go:55 +0x225
main.init.0()
        /home/dearr/Work/whistle/backend/main.go:43 +0xb8
exit status 2
```

my go.mod has this versions:
```go
github.com/casbin/casbin-pg-adapter v1.0.2
github.com/casbin/casbin/v2 v2.30.0
```

--- a/adapter.go
+++ b/adapter.go
@@ -471,30 +471,119 @@ func (a *Adapter) UpdatePolicies(sec string, ptype string, oldRules, newRules []
 	return a.updatePolicies(oldLines, newLines)
 }
 
-func (a *Adapter) UpdateFilteredPolicies(sec string, ptype string, oldRules, newRules [][]string) error {
-	var oldLines []*CasbinRule
-	for _, rule := range oldRules {
-		line := savePolicyLine(ptype, rule)
-		oldLines = append(oldLines, line)
+func (a *Adapter) UpdateFilteredPolicies(sec string, ptype string, newPolicies [][]string, fieldIndex int, fieldValues ...string) ([][]string, error) {
+	line := &CasbinRule{}
+
+	line.PType = ptype
+	if fieldIndex <= 0 && 0 < fieldIndex+len(fieldValues) {
+		line.V0 = fieldValues[0-fieldIndex]
+	}
+	if fieldIndex <= 1 && 1 < fieldIndex+len(fieldValues) {
+		line.V1 = fieldValues[1-fieldIndex]
+	}
+	if fieldIndex <= 2 && 2 < fieldIndex+len(fieldValues) {
+		line.V2 = fieldValues[2-fieldIndex]
+	}
+	if fieldIndex <= 3 && 3 < fieldIndex+len(fieldValues) {
+		line.V3 = fieldValues[3-fieldIndex]
+	}
+	if fieldIndex <= 4 && 4 < fieldIndex+len(fieldValues) {
+		line.V4 = fieldValues[4-fieldIndex]
+	}
+	if fieldIndex <= 5 && 5 < fieldIndex+len(fieldValues) {
+		line.V5 = fieldValues[5-fieldIndex]
 	}
 
-	var newLines []*CasbinRule
-	for _, rule := range newRules {
-		line := savePolicyLine(ptype, rule)
-		newLines = append(newLines, line)
+	newP := make([]CasbinRule, 0, len(newPolicies))
+	oldP := make([]CasbinRule, 0)
+	for _, newRule := range newPolicies {
+		newP = append(newP, *(savePolicyLine(ptype, newRule)))
 	}
 
 	err := a.db.RunInTransaction(context.Background(), func(tx *pg.Tx) error {
-		_, err := tx.Model(&oldLines).Table(a.tableName).Delete()
-		if err != nil {
-			return err
+		for i := range newP {
+			str, args := line.queryString()
+			_, err := tx.Model(&oldP).Table(a.tableName).Where(str, args...).Delete()
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+			_, err = tx.Model(&newP[i]).Table(a.tableName).
+				OnConflict("DO NOTHING").
+				Insert()
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
 		}
-		_, err = tx.Model(&newLines).Table(a.tableName).
-			OnConflict("DO NOTHING").
-			Insert()
-		return err
+		return nil
 	})
-	return err
+
+	// return deleted rulues
+	oldPolicies := make([][]string, 0)
+	for _, v := range oldP {
+		oldPolicy := v.toStringPolicy()
+		oldPolicies = append(oldPolicies, oldPolicy)
+	}
+	return oldPolicies, err
+}
+
+func (c *CasbinRule) queryString() (string, []interface{}) {
+	queryArgs := []interface{}{c.PType}
+
+	queryStr := "p_type = ?"
+	if c.V0 != "" {
+		queryStr += " and v0 = ?"
+		queryArgs = append(queryArgs, c.V0)
+	}
+	if c.V1 != "" {
+		queryStr += " and v1 = ?"
+		queryArgs = append(queryArgs, c.V1)
+	}
+	if c.V2 != "" {
+		queryStr += " and v2 = ?"
+		queryArgs = append(queryArgs, c.V2)
+	}
+	if c.V3 != "" {
+		queryStr += " and v3 = ?"
+		queryArgs = append(queryArgs, c.V3)
+	}
+	if c.V4 != "" {
+		queryStr += " and v4 = ?"
+		queryArgs = append(queryArgs, c.V4)
+	}
+	if c.V5 != "" {
+		queryStr += " and v5 = ?"
+		queryArgs = append(queryArgs, c.V5)
+	}
+
+	return queryStr, queryArgs
+}
+
+func (c *CasbinRule) toStringPolicy() []string {
+	policy := make([]string, 0)
+	if c.PType != "" {
+		policy = append(policy, c.PType)
+	}
+	if c.V0 != "" {
+		policy = append(policy, c.V0)
+	}
+	if c.V1 != "" {
+		policy = append(policy, c.V1)
+	}
+	if c.V2 != "" {
+		policy = append(policy, c.V2)
+	}
+	if c.V3 != "" {
+		policy = append(policy, c.V3)
+	}
+	if c.V4 != "" {
+		policy = append(policy, c.V4)
+	}
+	if c.V5 != "" {
+		policy = append(policy, c.V5)
+	}
+	return policy
 }
 
 func (a *Adapter) updatePolicies(oldLines, newLines []*CasbinRule) error {
@@ -505,8 +594,8 @@ func (a *Adapter) updatePolicies(oldLines, newLines []*CasbinRule) error {
 	defer tx.Close()
 
 	for i, line := range oldLines {
-		newLines[i].ID = line.ID
-		_, err = tx.Model(newLines[i]).Table(a.tableName).WherePK().Update()
+		str, args := line.queryString()
+		_, err = tx.Model(newLines[i]).Table(a.tableName).Where(str, args...).Update()
 		if err != nil {
 			tx.Rollback()
 			return err

--- a/adapter_test.go
+++ b/adapter_test.go
@@ -335,7 +335,9 @@ func (s *AdapterTestSuite) TestUpdateFilteredPolicies() {
 	err = s.e.SavePolicy()
 	s.Require().NoError(err)
 
-	err = s.a.UpdateFilteredPolicies("p", "p", [][]string{{"alice", "data1", "read"}, {"bob", "data2", "write"}}, [][]string{{"alice", "data2", "write"}, {"bob", "data1", "read"}})
+	_, err = s.a.UpdateFilteredPolicies("p", "p", [][]string{{"alice", "data2", "write"}}, 0, "alice", "data1", "read")
+	s.Require().NoError(err)
+	_, err = s.a.UpdateFilteredPolicies("p", "p", [][]string{{"bob", "data1", "read"}}, 0, "bob", "data2", "write")
 	s.Require().NoError(err)
 
 	err = s.e.LoadPolicy()

--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,7 @@ module github.com/casbin/casbin-pg-adapter
 go 1.14
 
 require (
-	github.com/casbin/casbin/v2 v2.23.4
+	github.com/casbin/casbin/v2 v2.36.1
 	github.com/go-pg/pg/v10 v10.9.1
 	github.com/mmcloughlin/meow v0.0.0-20181112033425-871e50784daf
 	github.com/stretchr/testify v1.7.0

